tools/python/search_index.py

```diff
@@ -0,0 +1,237 @@
+import os
+import re
+import json
+import logging
+import subprocess
+
+from html.parser import HTMLParser
+
+try:
+    from lunr import lunr
+    haslunrpy = True
+except ImportError:
+    haslunrpy = False
+
+log = logging.getLogger(__name__)
+
+
+class SearchIndex:
+    """
+    Search index is a collection of pages and sections (heading
+    tags and their following content are sections).
+    """
+
+    def __init__(self, **config):
+        self._entries = []
+        self.config = config
+
+    def _find_toc_by_id(self, toc, id_):
+        """
+        Given a table of contents and HTML ID, iterate through
+        and return the matched item in the TOC.
+        """
+        for toc_item in toc:
+            if toc_item.id == id_:
+                return toc_item
+            toc_item_r = self._find_toc_by_id(toc_item.children, id_)
+            if toc_item_r is not None:
+                return toc_item_r
+
+    def _add_entry(self, title, text, loc):
+        """
+        A simple wrapper to add an entry, dropping bad characters.
+        """
+        text = text.replace('\u00a0', ' ')
+        text = re.sub(r'[ \t\n\r\f\v]+', ' ', text.strip())
+
+        self._entries.append({
+            'title': title,
+            'text': text,
+            'location': loc
+        })
+
+    def add_entry_from_context(self, page):
+        """
+        Create a set of entries in the index for a page. One for
+        the page itself and then one for each of its' heading
+        tags.
+        """
+
+        # Create the content parser and feed in the HTML for the
+        # full page. This handles all the parsing and prepares
+        # us to iterate through it.
+        parser = ContentParser()
+
+        # Fix the spacing around hightlighted content in code blocks
+        # i.e. removing space before and after </span> tag
+        tag_regex = re.compile('\s*<span([^>]*)>')
+        html = tag_regex.sub('', page.content)
+        html = re.sub('</span>\s*','', html)
```

Comment:
correct issue with spaces in code blocks
```suggestion
        tag_regex = re.compile('\s*<span([^>]*)>')
        tmp_page = re.sub('<span class="w">\s*','&nbsp;', page.content)
        html = tag_regex.sub('', tmp_page)
        html = re.sub('</span>\s*','', html)
```